### PR TITLE
Log queries in the background

### LINKED ISSUE #249: Workers should not be responsible for data logging
**Is your feature request related to a problem? Please describe:**  
Currently, every worker is logging the queries it has executed in the last second by itself. That has the disadvantage, that the worker isn't executing the queries while logging. Based on the measurements, logging can last up to 500ms, which means, that the worker is wasting up to 50% of the time by logging instead of query executing - its primary task. Moreover, because InfluxDB is single-threaded, the workers compete for Influx access, slowing down each other.

**Please describe the solution you'd like:**  
There are some ideas about how the problem could be solved (additional solutions are welcome):
(Influx data set was cleared before running each experiment)

### Solution 1 
Add a multiprocessing logging queue, where workers store logging information about the queries they have executed. Define an `appsheduler` job, which runs every second, clears the queue and logs its content to the Influx.

This idea was tested with Python script and gives the following numbers:
<details><summary>Code</summary>
<p>

```python
from influxdb import InfluxDBClient
from time import time, sleep
from multiprocessing import Process, Queue
from apscheduler.schedulers.background import BackgroundScheduler
import psycopg2
import csv

client = InfluxDBClient('localhost', '8086', 'root', 'root')
logs = []
log_counter = 0

def execute_queries(worker_id, queue, limit, n_workers, table_size):
    queue.cancel_join_thread()
    log = []
    conn = psycopg2.connect(dbname='postgres', user='serviceuser', host='vm-york.eaalab.hpi.uni-potsdam.de', password='serviceuser')
    conn.set_session(autocommit=True)
    cur = conn.cursor()
    counter = 0
    while counter < limit:
        start = time()
        cur.execute('SELECT 1;')
        end = time()

        push_start = time()
        queue.put((start, end, worker_id, counter))
        push_end = time()

        log.append([push_start, push_end - push_start, counter, worker_id])
        counter = counter + 1

    with open(f"Samples/Queue_push_{table_size}_{n_processes}_w{worker_id}.csv", "w", newline="") as f:
        filednames = ["time", "duration", "query_no", "worker_id"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(log)


def log_data(queue):
    global log_counter
    global logs
    measurements = []
    queue_start = time()
    while not queue.empty():
        measurements.append(queue.get())
    queue_end = time()
    points = []
    for i in range(len(measurements)):
        points.append({
                    "measurement": "successful_queries",
                    "tags": {"worker_id": measurements[i][3], "query_no": measurements[i][2]},
                    "fields": {"start": measurements[i][0], "end": measurements[i][1]},
                    "time": log_counter,
                })
        log_counter = log_counter + 1
    write_start = time()
    if(len(points) > 0):
        client.write_points(points, database="york")
    write_end = time()
    size = len(points)
    print(f"""Wrote {size} points""")
    logs.append([queue_start, queue_end - queue_start, write_end - write_start, len(measurements)])


def benchmark(n_workers, table_size):
    global logs
    logs = []
    processes = []
    counter = 0
    queue = Queue(0)
    scheduler = BackgroundScheduler()
    log_data_job = scheduler.add_job(
                func=log_data, args=(queue,), trigger="interval", seconds=1,
            )
    print("Preparation...")
    for i in range(n_workers):
        p = Process(target=execute_queries, args=(i, queue, table_size/n_workers, n_workers, table_size))
        processes.append(p)
    print("Starting...")
    for i in range(n_processes):
        processes[i].start()
    scheduler.start()
    print("Waiting for join...")
    for i in range(n_processes):
        processes[i].join()
    print("Waiting for scheduler...")
    size = queue.qsize()
    print(f"Qsize: {size}")
    log_data_job.remove()
    scheduler.shutdown()

def write_results(n_processes):
    print("Printing...")
    with open(f"Samples/Queue_get_{table_size}_{n_processes}.csv", "w", newline="") as f:
        filednames = ["time", "get duration", "write duration", "number_queries"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(logs)


n_processes = 10
table_size = 1000000
print(f"Using {n_processes} processes. Table size: {table_size}")

benchmark(n_processes, table_size)
write_results(n_processes)

print("Finished")

```

</p>
</details>

image
The average duration of inserting one item to the logging queue: 0.011ms

As we see, in that case, the clearing of the queue can last up to 1 sec, which can be too slow.

### Solution 2
Same idea as in solution 1, but define logging queue for every worker. The `appscheduler` job should clear all queues and log its content to Influx. In that case, we could reduce the number of concurrent writes to the queue.

<details><summary>Code</summary>
<p>

```python
from influxdb import InfluxDBClient
from time import time, sleep
from multiprocessing import Process, Queue
from apscheduler.schedulers.background import BackgroundScheduler
import psycopg2
import csv

client = InfluxDBClient('localhost', '8086', 'root', 'root')
logs = []
log_counter = 0

def execute_queries(worker_id, queue, limit, n_workers, table_size):
    queue.cancel_join_thread()
    log = []
    conn = psycopg2.connect(dbname='postgres', user='serviceuser', host='vm-york.eaalab.hpi.uni-potsdam.de', password='serviceuser')
    conn.set_session(autocommit=True)
    cur = conn.cursor()
    counter = 0
    while counter < limit:
        start = time()
        cur.execute('SELECT 1;')
        end = time()

        push_start = time()
        queue.put((start, end, worker_id, counter))
        push_end = time()

        log.append([push_start, push_end - push_start, counter, worker_id])
        counter = counter + 1

    with open(f"Samples/ManyQueues_push_{table_size}_{n_processes}_w{worker_id}.csv", "w", newline="") as f:
        filednames = ["time", "duration", "query_no", "worker_id"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(log)


def log_data(queues, n_workers):
    global log_counter
    global logs
    measurements = []
    queue_start = time()
    for i in range(n_workers):
        while not queues[i].empty():
            measurements.append(queues[i].get())
    queue_end = time()
    points = []
    for i in range(len(measurements)):
        points.append({
                    "measurement": "successful_queries",
                    "tags": {"worker_id": measurements[i][3], "query_no": measurements[i][2]},
                    "fields": {"start": measurements[i][0], "end": measurements[i][1]},
                    "time": log_counter,
                })
        log_counter = log_counter + 1
    write_start = time()
    if(len(points) > 0):
        client.write_points(points, database="york")
    size = len(points)
    print(f"""Wrote {size} points""")
    write_end = time()
    logs.append([queue_start, queue_end - queue_start, write_end - write_start, len(measurements)])


def benchmark(n_workers, table_size):
    global logs
    logs = []
    processes = []
    counter = 0
    queues = []
    scheduler = BackgroundScheduler()
    print("Preparation...")
    for i in range(n_workers):
        queues.append(Queue(0))
        p = Process(target=execute_queries, args=(i, queues[i], table_size/n_workers, n_workers, table_size))
        processes.append(p)
    log_data_job = scheduler.add_job(
                func=log_data, args=(queues,n_workers), trigger="interval", seconds=1,
            )
    print("Starting...")
    for i in range(n_processes):
        processes[i].start()
    scheduler.start()
    print("Waiting for join...")
    for i in range(n_processes):
        processes[i].join()
    print("Waiting for scheduler...")
    for i in range(n_workers):
        size = queues[i].qsize()
        print(f"Qsize: {size}")
    log_data_job.remove()
    scheduler.shutdown()

def write_results(n_processes):
    print("Printing...")
    with open(f"Samples/ManyQueues_get_{table_size}_{n_processes}.csv", "w", newline="") as f:
        filednames = ["time", "get duration", "write duration", "number_queries"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(logs)


n_processes = 10
table_size = 1000000
print(f"Using {n_processes} processes. Table size: {table_size}")

benchmark(n_processes, table_size)
write_results(n_processes)

print("Finished")

```

</p>
</details>

image
The average duration of inserting one item to the logging queue: 0.014ms
Despite expectations (there were no concurrent writes to the queue this time), this metric has increased.

The total execution time of the job looks a bit better, but it still overrides the limit of 1 sec 

### Solution 3
Use one queue for logging, but define a Process, which clears queue and logs query information continuously and sleeps 200ms, if the queue is empty.   

<details><summary>Code</summary>
<p>

```python
from influxdb import InfluxDBClient
from time import time, sleep
from multiprocessing import Process, Queue, Value
from apscheduler.schedulers.background import BackgroundScheduler
import psycopg2
import csv

client = InfluxDBClient('localhost', '8086', 'root', 'root')
logs = []
log_counter = 0

def execute_queries(worker_id, queue, limit, n_workers, table_size):
    queue.cancel_join_thread()
    log = []
    conn = psycopg2.connect(dbname='postgres', user='serviceuser', host='vm-york.eaalab.hpi.uni-potsdam.de', password='serviceuser')
    conn.set_session(autocommit=True)
    cur = conn.cursor()
    counter = 0
    while counter < limit:
        start = time()
        cur.execute('SELECT 1;')
        end = time()

        push_start = time()
        queue.put((start, end, worker_id, counter))
        push_end = time()

        log.append([push_start, push_end - push_start, counter, worker_id])
        counter = counter + 1

    with open(f"Samples/ProcessQueue_push_{table_size}_{n_processes}_w{worker_id}.csv", "w", newline="") as f:
        filednames = ["time", "duration", "query_no", "worker_id"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(log)


def log_data(queue, stop_flag, table_size, n_processes):
    log_counter = 0
    logs = []
    while True:
        if stop_flag.value:
            break
        measurements = []
        queue_start = time()
        while not queue.empty():
            measurements.append(queue.get())
        queue_end = time()
        points = []
        for i in range(len(measurements)):
            points.append({
                        "measurement": "successful_queries",
                        "tags": {"worker_id": measurements[i][3], "query_no": measurements[i][2]},
                        "fields": {"start": measurements[i][0], "end": measurements[i][1]},
                        "time": log_counter,
                    })
            log_counter = log_counter + 1
        write_start = time()
        if(len(points) > 0):
            client.write_points(points, database="york")
        write_end = time()
        size = len(points)
        print(f"""Wrote {size} points""")
        logs.append([queue_start, queue_end - queue_start, write_end - write_start, len(measurements)])
        if (len(measurements) == 0):
            sleep(0.2)

    with open(f"Samples/ProcessQueue_get_{table_size}_{n_processes}.csv", "w", newline="") as f:
        filednames = ["time", "get duration", "write duration", "number_queries"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(logs)

def benchmark(n_workers, table_size):
    processes = []
    counter = 0
    queue = Queue(0)
    print("Preparation...")
    stop_flag = Value('b', False)
    for i in range(n_workers):
        p = Process(target=execute_queries, args=(i, queue, table_size/n_workers, n_workers, table_size))
        processes.append(p)
    updater = Process(target=log_data, args=(queue, stop_flag, table_size, n_workers))
    print("Starting...")
    for i in range(n_processes):
        processes[i].start()
    updater.start()
    print("Waiting for join...")
    for i in range(n_processes):
        processes[i].join()
    stop_flag.value = True
    size = queue.qsize()
    print(f"Qsize: {size}")
    print("Waiting for updater...")
    updater.join()

n_processes = 10
table_size = 100000
print(f"Using {n_processes} processes. Table size: {table_size}")

benchmark(n_processes, table_size)

print("Finished")

```

</p>
</details>

image
The average duration of inserting one item to the logging queue: 0.013ms

This looks way more performant, but the process of queue clearing takes almost as much time as a write operation.

### Solution 4
Use one queue for logging, but define a Process, which clears queue and logs query information continuously and sleeps 200ms, if the queue is empty. The workers write query information to the logging queue in the bunches every second.

<details><summary>Code</summary>
<p>

```python
from influxdb import InfluxDBClient
from time import time, sleep
from multiprocessing import Process, Queue, Value
from apscheduler.schedulers.background import BackgroundScheduler
import psycopg2
import csv

client = InfluxDBClient('localhost', '8086', 'root', 'root')
logs = []
log_counter = 0

def execute_queries(worker_id, queue, limit, n_workers, table_size):
    log = []
    conn = psycopg2.connect(dbname='postgres', user='serviceuser', host='vm-york.eaalab.hpi.uni-potsdam.de', password='serviceuser')
    conn.set_session(autocommit=True)
    cur = conn.cursor()
    counter = 0
    last_logged = time()
    measurements = []
    while counter < limit:
        start = time()
        cur.execute('SELECT 1;')
        end = time()
        measurements.append((start, end, worker_id, counter))
        if (time() - last_logged > 1.0):
            push_start = time()
            queue.put(measurements)
            push_end = time()
            last_logged = push_end
            measurements = []
            log.append([push_start, push_end - push_start, counter, worker_id])
        counter = counter + 1
    queue.close()
    with open(f"Samples/Queue_push_{table_size}_{n_processes}_w{worker_id}.csv", "w", newline="") as f:
        filednames = ["time", "duration", "query_no", "worker_id"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(log)
    print(f"worker {worker_id} done")


def log_data(queue, stop_flag, table_size, n_processes):
    log_counter = 0
    logs = []
    while True:
        if stop_flag.value:
            break
        measurements = []
        queue_start = time()
        while not queue.empty():
            current = queue.get()
            for el in current:
                measurements.append(el)
        queue_end = time()
        points = []
        for i in range(len(measurements)):
            points.append({
                        "measurement": "successful_queries",
                        "tags": {"worker_id": measurements[i][3], "query_no": measurements[i][2]},
                        "fields": {"start": measurements[i][0], "end": measurements[i][1]},
                        "time": log_counter,
                    })
            log_counter = log_counter + 1
        write_start = time()
        if(len(points) > 0):
            client.write_points(points, database="york")
        write_end = time()
        size = len(points)
        print(f"""Wrote {size} points""")
        if size > 0:
            logs.append([queue_start, queue_end - queue_start, write_end - write_start, len(measurements)])
        if (len(measurements) == 0):
            sleep(0.2)

    with open(f"Samples/Queue_get_{table_size}_{n_processes}.csv", "w", newline="") as f:
        filednames = ["time", "get duration", "write duration", "number_queries"]
        csv_writer = csv.writer(f, delimiter ='|')
        csv_writer.writerow(filednames)
        csv_writer.writerows(logs)

def benchmark(n_workers, table_size):
    processes = []
    counter = 0
    queue = Queue(0)
    queue.cancel_join_thread()
    print("Preparation...")
    stop_flag = Value('b', False)
    for i in range(n_workers):
        p = Process(target=execute_queries, args=(i, queue, table_size/n_workers, n_workers, table_size))
        processes.append(p)
    updater = Process(target=log_data, args=(queue, stop_flag, table_size, n_workers))
    print("Starting...")
    for i in range(n_processes):
        processes[i].start()
    updater.start()
    print("Waiting for join...")
    for i in range(n_processes):
        processes[i].join()
    stop_flag.value = True
    size = queue.qsize()
    print(f"Qsize: {size}")
    print("Waiting for updater...")
    updater.join()

n_processes = 10
table_size = 1000000
print(f"Using {n_processes} processes. Table size: {table_size}")

benchmark(n_processes, table_size)

print("Finished")

```

</p>
</details>

image
The average duration of inserting one item to the logging queue: 0.042ms

**Affected Component(s):**  
`DatabaseManager`, `Influx`

**Additional context:**  
We need to ensure, that logging is fast enough to provide reliable throughput and latency values when calling corresponding endpoints in the backend.
Related to #202.

### PR DESCRIPTION
**Pull request by PeterTsayun**
# Resolves #249

**Does your pull request solve a problem? Please describe:**  
Until now, every second every worker needed to log successful queries. In the meantime, it couldn't execute queries, which led to a reduced load. Now, workers start a Thread to log the queries (network operation).

**Does your pull request add a feature? Please describe:**  
Log successful queries in the background.

**Affected Component(s):**  
`Worker`

**Additional context:**

### CODE REVIEW

File: hyrisecockpit/database_manager/worker.py [hunk 1]
```diff
@@ -133,6 +128,14 @@ def execute_queries(
                 ) as e:
                     failed_queries.append((time_ns(), worker_id, str(task), str(e)))
 
-                last_batched = log_results(
-                    log, last_batched, succesful_queries, failed_queries
-                )
+                if last_batched < time_ns() - 1_000_000_000 and (
+                    succesful_queries != [] or failed_queries != []
```
Comment (heseding) [inline hyrisecockpit/database_manager/worker.py]: https://docs.python.org/3.9/library/stdtypes.html#truth-value-testing
```suggestion
                    succesful_queries or failed_queries
```

File: hyrisecockpit/database_manager/worker.py [hunk 2]
```diff
@@ -74,18 +75,12 @@ def execute_task(
 
 def log_results(
     log: StorageCursor,
-    last_batched: int,
     succesful_queries: List[Tuple[int, int, str, str, str]],
     failed_queries: List[Tuple[int, str, str, str]],
-) -> int:
+) -> None:
     """Log results to database."""
-    if last_batched < time_ns() - 1_000_000_000:
```
Comment (heseding) [inline hyrisecockpit/database_manager/worker.py:82]: I think it's easier to understand the logging time check here.

**Review by heseding — APPROVED** [review]

File: hyrisecockpit/database_manager/worker.py (hunk 2 above)
Comment (PeterTsayun) [reply hyrisecockpit/database_manager/worker.py:82]: Wouldn't it imply, that in that case, we would create a thread after every single query execution, regardless of the fact, whether we want to log the queries or not?

**Review by Alexander-Dubrawski — APPROVED** [review]

File: hyrisecockpit/database_manager/worker.py (hunk 2 above)
Comment (heseding) [reply hyrisecockpit/database_manager/worker.py:82]: Yes, I must've missed that.

**Pull request merged**